UM-USElab/gradecraft-development
# Auto award unlocks

### Status
**READY**

### Description
This is a large feature branch PR that involves unlocks, auto awarding badges, and email notifications. 
Logic was added to the models for grade, submission, earned_badge, and course membership to check for auto awarded badges and send email after a change to those models triggers an unlock. 
unlockable_condition concern was changed to yield the unlock_state after the conditions are met. 
Auto award on unlock concern was created as a place to add methods that help facilitate any automatic actions taking place after something has been unlocked. 
Spec tests were added for the auto awarding of badges, one was corrected for the changes we made to the unlockable_condition concerns unlock! method 

### Todos
- [ ] Possibly add tests around automatically sending emails?
- [ ] User testing for badge conditions and auto awarding sed badge

### Deploy Notes
Made migration, first migration made since upgrade to new version lots of additions to schema.rb
Added .byebug* to .gitignore

### Migrations
YES -- made migration to add an auto-award boolean to the badges model

### Steps to Test or Reproduce
create new badge or change existing badge to have 'auto-award' column checked on badges/36/edit at the unlocks condition tab under visibility settings then add one ore more condition requirements to unlock the badge. 

Complete the requirements set for the badge, then a badge should be automatically awarded and emails sent. To test emails locally start an smtp like mailcatcher and change development.rb logs config.action_mailer.perform_deliveries = false 

### Impacted Areas in Application
Grade, course_membership, earned_badge, submissions, email notifications 

======================
Closes #1998
